### ISSUE TITLE
The LayerDrop training scripts

### ISSUE THREAD

**Issue by sunmeng01** [pytorch/fairseq#3164]
@huihuifan I have try many experiments on En-De WMT 14 to reproduce the SOTA bleu score 30.2 by big Transformer with 12-encoders, 6-decoders and layerdrop.
exp1:
python train.py $data_dir \
    --save-dir $save_dir \
    --arch transformer_vaswani_wmt_en_de_big  \
    --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
    --lr 0.001 --update-freq 16 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 \
    --dropout 0.3 --weight-decay 0.0 --encoder-layers 12 --decoder-layers 6 \
    --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
    --max-tokens 3584 \
    --fp16 --encoder-layerdrop 0.3 --decoder-layerdrop 0.3 --ddp-backend no_c10d

average checkpoints from epoch 61-70, get the best bleu 29.46

exp2:
python train.py $data_dir \
    --save-dir $save_dir \
    --arch transformer_vaswani_wmt_en_de_big  \
    --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
    --lr 0.001 --update-freq 16 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 \
    --dropout 0.3 --weight-decay 0.0 --encoder-layers 12 --decoder-layers 6 \
    --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
    --max-tokens 3584 \
    --fp16 --encoder-layerdrop 0.0 --decoder-layerdrop 0.0 --ddp-backend no_c10d

 average checkpoints from epoch 60-69, get the best bleu 29.64

exp3:
python train.py $data_dir \
    --save-dir $save_dir \
    --arch transformer_vaswani_wmt_en_de_big  \
    --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
    --lr 0.001 --update-freq 16 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 \
    --dropout 0.3 --weight-decay 0.0 --encoder-layers 12 --decoder-layers 6 \
    --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
    --max-tokens 3584 \
    --fp16 --encoder-layerdrop 0.2 --decoder-layerdrop 0.2 --ddp-backend no_c10d

average checkpoints from epoch 60-69, get the best bleu 29.64

exp4:
python train.py $data_dir \
    --save-dir $save_dir \
    --arch transformer_vaswani_wmt_en_de_big  \
    --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
    --lr 0.001 --update-freq 16 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 \
    --dropout 0.2 --weight-decay 0.0 --encoder-layers 12 --decoder-layers 6 \
    --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
    --max-tokens 3584 \
    --fp16 --encoder-layerdrop 0.1 --decoder-layerdrop 0.2 --ddp-backend no_c10d 

average checkpoints from epoch 57-66, get the best bleu 29.19

exp5:
python train.py $data_dir \
    --save-dir $save_dir \
    --arch transformer_vaswani_wmt_en_de_big  \
    --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
    --lr 0.001 --update-freq 16 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 \
    --dropout 0.1 --weight-decay 0.0 --encoder-layers 12 --decoder-layers 6 \
    --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
    --max-tokens 4096 \
    --fp16 --encoder-layerdrop 0.1 --encoder-layerdrop 0.2 --ddp-backend no_c10d

average checkpoints from epoch 64-73, get the best bleu 29.41

In the experiments, I try vary dropout from 0.1-0.3,  encoder-layerdrop and encoder-layerdrop from 0.0-0.3, however, I cannot get the SOTA bleu 30.2 in the LayerDrop paper.
The decoding parameters  are --beam 4 --batch-size 100 --remove-bpe --lenpen 0.6
And also use the post script "perl -ple 's{(\S)-(\S)}{$1 ##AT##-##AT## $2}g'" for reference and translation
In the decoding phase, I just use 12 encoders and 6 decoders without pruning any layer.

Would you please provide the training scripts for me to get bleu 30.2?
Thanks~

**Comment by huihuifan** [comment]
hello, thanks for opening this task. are you applying compound splitting? https://github.com/pytorch/fairseq/blob/master/scripts/compound_split_bleu.sh
unfortunately, the original transformer paper uses this, so we used it too to have comparable BLEU. there's some details in examples/scaling_nmt about this. ideally we should report detokenized bleu with sacrebleu only probably.

**Comment by sunmeng01** [comment]
Thanks for your reply. Yes, all the BLEU score reported above are calculated by compound_split_bleu.sh. And I can get baseline bleu 29.3 following examples/scaling_nmt

**Comment by huihuifan** [comment]
can you try with a different beam size? 

fairseq-generate /path/to/bped/wmt/data --path nmt_checkpoint.pt \
  --beam 8 --lenpen 0.4 \
  --batch-size 64 \
  --remove-bpe \
  --gen-subset test > wmt16_gen.txt
bash scripts/compound_split_bleu.sh wmt16_gen.txt
# prints BLEU4 = 30.17

looks like what i used in layerdrop. i believe i copied the beam 8 from the hassan et al paper about chinese to english translation

**Comment by sunmeng01** [comment]
Thanks~ I will try beam 8~

**Comment by netw0rkf10w** [comment]
Hi @huihuifan,

Could you please tell me how many steps did you train your model for? 
Intuitively LayerDrop seems to require more training steps to surpass the [baseline](https://arxiv.org/abs/1806.00187) (similar to dropout), but I couldn't find this implementation detail in either papers or in the [documentation](https://github.com/pytorch/fairseq/blob/master/examples/scaling_nmt/README.md).

I tried training the model (without LayerDrop) on WMT16 on 16 GPUs for a day (>150k steps) but obtained only 27.2 SacreBLEU.

Thank you in advance for your reply!
